src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -1186,140 +1081,130 @@ private string GetLogicForDefaultSerializerOptionsInit()
-
-        if (converter.CanConvert(type))
-        {{
-            if (converter is {JsonConverterFactoryTypeRef} factory)
-            {{
-                converter = factory.CreateConverter(type, {OptionsLocalVariableName});
-                if (converter == null || converter is {JsonConverterFactoryTypeRef})
-                {{
-                    throw new {InvalidOperationExceptionTypeRef}(string.Format(""{ExceptionMessages.InvalidJsonConverterFactoryOutput}"", factory.GetType()));
-                }}
-            }}
-
-            return converter;
-        }}
-    }}
-
-    return null;
-}}";
-            }
+                // The generic type parameter could capture type parameters from containing types,
+                // so use a name that is unlikely to be used.
+                const string TypeParameter = "TJsonMetadataType";
 
-            private static string GetFetchLogicForGetCustomConverter_PropertiesWithFactories()
-            {
-                return @$"
+                writer.WriteLine($$"""
+                    private static bool {{TryGetTypeInfoForRuntimeCustomConverterMethodName}}<{{TypeParameter}}>({{JsonSerializerOptionsTypeRef}} options, out {{JsonTypeInfoTypeRef}}<{{TypeParameter}}> jsonTypeInfo)
+                    {
+                        {{JsonConverterTypeRef}}? converter = GetRuntimeConverterForType(typeof({{TypeParameter}}), options);
+                        if (converter != null)
+                        {
+                            jsonTypeInfo = {{JsonMetadataServicesTypeRef}}.{{CreateValueInfoMethodName}}<{{TypeParameter}}>(options, converter);
+                            return true;
+                        }
 
-private static {JsonConverterTypeRef}<T> {GetConverterFromFactoryMethodName}<T>({JsonSerializerOptionsTypeRef} {OptionsLocalVariableName}, {JsonConverterFactoryTypeRef} factory)
-{{
-    return ({JsonConverterTypeRef}<T>) {GetConverterFromFactoryMethodName}({OptionsLocalVariableName}, typeof(T), factory);
-}}";
-            }
+                        jsonTypeInfo = null;
+                        return false;
+                    }
 
-            private static string GetFetchLogicForGetCustomConverter_TypesWithFactories()
-            {
-                return @$"
-
-private static {JsonConverterTypeRef} {GetConverterFromFactoryMethodName}({JsonSerializerOptionsTypeRef} {OptionsLocalVariableName}, {TypeTypeRef} type, {JsonConverterFactoryTypeRef} factory)
-{{
-    {JsonConverterTypeRef}? converter = factory.CreateConverter(type, {OptionsLocalVariableName});
-    if (converter == null || converter is {JsonConverterFactoryTypeRef})
-    {{
-        throw new {InvalidOperationExceptionTypeRef}(string.Format(""{ExceptionMessages.InvalidJsonConverterFactoryOutput}"", factory.GetType()));
-    }}
-
-    return converter;
-}}";
+                    private static {{JsonConverterTypeRef}}? GetRuntimeConverterForType({{TypeTypeRef}} type, {{JsonSerializerOptionsTypeRef}} options)
```

Comment:
Mostly out of curiosity, but I was wondering - for all these various private helpers we're generating, if they're specifically meant to save binary size (as they're essentially doing outlining of code that used to just be emitted in every metadata info accessor), should we consider adding `[MethodImpl(MethodImplOptions.NoInlining)]` on them to make sure the compiler never actually tries to inline them back? Again, if the intent here is specifically to save binary size and not speed 🤔